### ISSUE TITLE
The model item passed into the dictionary is of type 'Sitecore.Mvc.Presentation.RenderingModel', but this dictionary requires a model item of type XXXX

### ISSUE THREAD

**Issue by Antonytm** [mikeedwards83/Glass.Mapper#229]
1) Create model that correspond some template: ModelX
2) Create view that uses this model @inherits GlassView<ModelX>
3) Set wrong datasource in presentation details for this view

Actual result: Page crash(The model item passed into the dictionary is of type 'Sitecore.Mvc.Presentation.RenderingModel', but this dictionary requires a model item of type XXXX) 
Expected result: empty view should be rendered.

**Comment by Antonytm** [comment]
I solved this issue by changing GetModelFromView processor by adding 
                if (model == null && modelType.GetInterface("IGlassBase") != null)
                {
                      model = Activator.CreateInstance(modelType);
                }
to return empty model to view, but model of required type. However I am not sure how to fix it in general case, when you don't know that all your objects are inherited from some interface.

**Comment by cardinal252** [comment]
This behaviour is standard to MVC, since you are indeed passing an incorrect model to the view. It appears that the model being passed in this instance is being controlled by the standard Sitecore processes (hence passing a standard RenderingModel instance).

It is my opinion that this is designed behaviour rather than a bug ?

**Comment by Antonytm** [comment]
Yes, it is standard behavior. But when Content Manager do something wrong, it this case for example remove item that is assigned as datasource to rendering view then page crashes in EE. And there is no option to fix this problem via EE. It is possible to change datasource from Content Editor by changing presentation details, but it is very inconvenient, especially when there are a lot of controls. So, it would be great to have some protection for this situation, e.g. return empty object of correct type.

**Comment by cardinal252** [comment]
In most respects, this should be handled with correct setting of a Datasource Template.

This I think would be the same with Sitecore's default behaviour if a strongly typed model was defined ?

I think it needs further discussion and investigation to see if this could be protected against without interfering with the default Sitecore behaviour.

**Comment by KayeeNL** [comment]
he @cardinal252 @mikeedwards83 , this issue is happening to me as well... the problem already occurs if you paste a random guid in the datasource field (which can be done of course, or the guid is the remains of a deleted item and remove links wasn't used), even though the Datasource location and datasourc e template  has been set on the rendering.

Think i will go and investigate for now the solution by @Antonytm and see if it works for my situation.

--Robbert

**Comment by aiorgu** [comment]
Same issue here

**Comment by jgsousa** [comment]
Incompatible items can be prevented through datasource templates, but this happens with broken links, unpublished datasources, and it is impossible to handle via razor or C# code. If the datasource doesn't exist the item should be ignored gracefully instead of leading to a 500 server error.

**Comment by cardinal252** [comment]
It is the case that we are letting Sitecore handle it in the default manner. The suggested fix above is that we create a model regardless - this assumes that ALL models should be handled by Glass which may not be the case?

**Comment by cardinal252** [comment]
It would simple enough to potentially write an example GetModel processor that could be enabled in a customer solution, but I don't know if we can give a permanent fix to it ?

**Comment by YggY** [comment]
> Incompatible items can be prevented through datasource templates, but this happens with broken links, unpublished datasources or a datasource that doesn't have that particular language version, and it is impossible to handle via razor or C# code. If the datasource item doesn't exist it should be ignored gracefully instead of leading to a 500 server error.

Wouldn't a broken link/unpublished datasource be replaced by the default behaviour of a datasource: using the Context Item as datasource?

**Comment by mwillebrands** [comment]
@mikeedwards83 currently Glass does not follow the same datasource resolving as that Sitecore does. 
Glass will check if a DataSource has been set, if true it tries to get the item and returns it even if it is null. Sitecore checks if the item is null, if so, fall back on the second condition.

Current Glass situation:
- RenderingItem.DataSource.HasValue -> Return this even if null
- RenderingItem.RenderingItem.DataSource.HasValue -> Return this even if null
- RenderingItem.Item != null -> return this even if null
- Return ContextItem

Proposed Glass situation: (Add a configuration setting to switch between behaviour)
- RenderingItem.DataSource.HasValue -> If not null, return
- RenderingItem.RenderingItem.DataSource.HasValue ->  If not null, return
- RenderingItem.Item != null ->  If not null, return
- Return ContextItem

I don't mind changing this logic, but i was wondering what the reason is behind the Glass implementation, before i propose a change in a pull request.
